File: src/2024/day2/2/main.go
package main

import (
	_ "embed"
	"fmt"
	"strconv"
	"strings"
)

//go:embed data.txt
var fileContents string

func isSequenceSafe(sequence []int) bool {
	isIncreasing := true
	isDecreasing := true

	for i := 0; i < len(sequence)-1; i++ {
		if sequence[i] >= sequence[i+1] {
			isIncreasing = false
		}
		if sequence[i] <= sequence[i+1] {
			isDecreasing = false
		}
	}

	if !(isIncreasing || isDecreasing) {
		return false
	}

	for i := 0; i < len(sequence)-1; i++ {
		diff := abs(sequence[i] - sequence[i+1])
		if diff < 1 || diff > 3 {
			return false
		}
	}

	return true
}

func isReportSafe(report string) bool {
	strLevels := strings.Fields(report)
	levels := make([]int, len(strLevels))

	for i, strLevel := range strLevels {
		level, err := strconv.Atoi(strLevel)
		if err != nil {
			return false
		}
		levels[i] = level
	}

	if isSequenceSafe(levels) {
		return true
	}

	for i := range levels {
		testLevels := make([]int, 0, len(levels)-1)
		testLevels = append(testLevels, levels[:i]...)
		testLevels = append(testLevels, levels[i+1:]...)

		if isSequenceSafe(testLevels) {
			return true
		}
	}

	return false
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func main() {
	lines := strings.Split(fileContents, "\n")

	safeReports := 0
	for _, line := range lines {
		if isReportSafe(line) {
			safeReports++
		}
	}

	fmt.Println(safeReports)
}

File: src/2024/day2/2/main.rs
use std::fs;
use std::path::Path;

fn is_sequence_safe(sequence: &[i32]) -> bool {
    let is_increasing = sequence.windows(2).all(|w| w[0] < w[1]);
    let is_decreasing = sequence.windows(2).all(|w| w[0] > w[1]);

    if !(is_increasing || is_decreasing) {
        return false;
    }

    sequence.windows(2).all(|w| {
        let diff = (w[0] - w[1]).abs();
        diff >= 1 && diff <= 3
    })
}

fn is_report_safe(report: &str) -> bool {
    // parse the report into levels
    let levels: Vec<i32> = report
        .split_whitespace()
        .map(|x| x.parse().unwrap())
        .collect();

    // check if the full sequence is safe
    if is_sequence_safe(&levels) {
        return true;
    }

    // try removing each level and check if the resulting sequence is safe
    for i in 0..levels.len() {
        let test_levels: Vec<i32> = levels
            .iter()
            .enumerate()
            .filter(|&(index, _)| index != i)
            .map(|(_, &val)| val)
            .collect();

        if is_sequence_safe(&test_levels) {
            return true;
        }
    }

    false
}

fn read_lines() -> Vec<String> {
    let current_file = Path::new(file!());
    let current_dir = current_file
        .parent()
        .expect("Failed to get parent directory");
    let file_path = current_dir.join("data.txt");

    fs::read_to_string(file_path)
        .expect("Failed to read file")
        .lines()
        .map(|line| line.to_string())
        .filter(|line| !line.is_empty())
        .collect()
}

fn main() {
    let lines = read_lines();

    let safe_reports = lines.iter().filter(|line| is_report_safe(line)).count();

    println!("{}", safe_reports);
}
